app/src/main/java/com/example/android/codelabs/paging/data/GithubRemoteMediator.kt

```diff
@@ -0,0 +1,166 @@
+        private val repoDatabase: RepoDatabase
+) : RemoteMediator<Int, Repo>() {
+
+    /**
+     * The paging library will call this method when there is no more data in the database to
+     * load from.
+     *
+     * For every item we retrieve we store the previous and next keys, so we know how to request
+     * more data, based on that specific item.
+     *
+     * state.pages holds the pages that were loaded until now and saved in the database.
+     *
+     * If the load type is REFRESH it means that we are in one of the following situations:
+     *  * it's the first time when the query was triggered and there was nothing saved yet in the
+     *  database
+     *  * a swipe to refresh was called
+     * If the load type is START, we need to rely on the first item available in the pages
+     * retrieved so far to compute they key for the previous page.
+     * If the load type is END, we need to rely on the last item available in the pages
+     * retrieved so far to compute they key for the next page.
+     */
+    override suspend fun load(loadType: LoadType, state: PagingState<Int, Repo>): MediatorResult {
+        val page = when (loadType) {
+            LoadType.REFRESH -> {
+                val remoteKeys = getRemoteKeyClosestToCurrentPosition(state)
+                remoteKeys?.nextKey?.minus(1) ?: GITHUB_STARTING_PAGE_INDEX
+            }
+            LoadType.START -> {
+                val remoteKeys = getRemoteKeyForFirstItem(state)
+                if (remoteKeys == null) {
+                    // The LoadType is START so some data was loaded before,
+                    // so we should have been able to get remote keys
+                    // If the remoteKeys are null, then we're an invalid state and we have a bug
+                    throw InvalidObjectException("Remote key and the prevKey should not be null")
+                }
+                // If the previous key is null, then we can't request more data
+                val prevKey = remoteKeys.prevKey
+                if (prevKey == null) {
+                    return MediatorResult.Success(endOfPaginationReached = false)
+                }
+                remoteKeys.prevKey
+            }
+            LoadType.END -> {
+                val remoteKeys = getRemoteKeyForLastItem(state)
+                if (remoteKeys?.nextKey == null) {
+                    // The LoadType is END, so some data was loaded before,
+                    // so we should have been able to get remote keys
+                    // If the remoteKeys are null, then we're an invalid state and we have a bug
+                    throw InvalidObjectException("Remote key should not be null for $loadType")
+                }
+                remoteKeys.nextKey
+            }
+        }
+
+        val apiQuery = query + IN_QUALIFIER
+
+        try {
+            val apiResponse = service.searchRepos(apiQuery, page, state.config.pageSize)
+
+            return if (apiResponse.isSuccessful) {
```

Comment:
Do we still gain anything by using the callback call adapter if we still need to check for `HttpException` and `IOException`? I feel like we should just use the suspend adapter from Retrofit, since that will likely be more relatable to users who are interested in using Coroutines. Sorry to flip back and forth on this, chris may disagree with me here though and in that case we can just keep this :D